src/cockpit/channels/filesystem.py

```diff
@@ -268,3 +287,226 @@ def do_close(self):
+            return True
+
+        return False
+
+    def report_initial_state(self, fd: Handle) -> None:
+        if self.flag_onlydir_error(fd):
+            return
+
+        self.fd = fd
+
+        entries = {''}
+        if self.fnmatch:
+            try:
+                entries.update(os.listdir(f'/proc/self/fd/{self.fd}'))
+                self.effective_fnmatch = self.fnmatch
+            except OSError:
+                # If we failed to get an initial list, then report nothing from now on
+                self.effective_fnmatch = ''
+
+        self.process_update({e for e in entries if self.interesting(e)}, reset=True)
+
+    def do_inotify_event(self, mask: InotifyEvent, cookie: int, rawname: 'bytes | None') -> None:
+        logger.debug('do_notify_event(%r, %r, %r)', mask, cookie, rawname)
+        name = (rawname or b'').decode(errors='surrogateescape')
+
+        self.schedule_update(name)
+
+        if name and mask | (InotifyEvent.CREATE | InotifyEvent.DELETE |
+                            InotifyEvent.MOVED_TO | InotifyEvent.MOVED_FROM):
+            # These events change the mtime of the directory
+            self.schedule_update('')
+
+    def do_identity_changed(self, fd: 'Handle | None', err: 'int | None') -> None:
+        # If there were previously pending changes, they are now irrelevant.
+        if self.pending is not None:
+            # Note: don't set to None, since the handler is still pending
+            self.pending.clear()
+
+        if err is None:
+            assert fd is not None
+            self.report_initial_state(fd)
+        else:
+            self.report_error(err)
+
+    def do_close(self) -> None:
+        if self.path_watch is not None:
+            self.path_watch.close()
+        self.close()
+
+    def do_open(self, options: JsonObject) -> None:
+        self.path = get_str(options, 'path')
+        if not os.path.isabs(self.path):
+            raise JsonError(options, '"path" must be an absolute path')
+
+        self.getattrs = self.make_getattrs(get_strv(options, 'attrs'))
+        self.fnmatch = get_str(options, 'fnmatch', '')
+        self.follow = get_bool(options, 'follow', default=True)
+        self.watch = get_bool(options, 'watch', default=False)
+        if self.watch and not self.follow:
+            raise JsonError(options, '"watch: true" and "follow: false" are (currently) incompatible')
+
+        self.current_value = {}
+        self.ready()
+
+        if not self.watch:
+            try:
+                fd = Handle.open(self.path, os.O_PATH if self.follow else os.O_PATH | os.O_NOFOLLOW)
+            except OSError as exc:
+                self.report_error(exc.errno)
+            else:
+                self.report_initial_state(fd)
+                fd.close()
+
+            self.done()
+            self.close()
+
+        else:
+            # PathWatch will essentially do the same as above
+            self.path_watch = PathWatch(self.path, self)
```

Comment:
Note to self: on errors, PathWatch calls `do_identity_changed(None, errno)`. 5 points for Lissindor if the comment says that :grin:

Reply:
Are you asking for more comments? For the five house points, I'd totally do it.

Reply:
Yes, pretty much what I  wrote, i.e. "will essentially do the same as above" → "calls do_identity_changed(None, errno) with the same behavior as above" or so

Reply:
I'd like my 5 points, please.

Reply:
⭐ ⭐ ⭐ ⭐ ⭐

Reply:
:heart: